Change update_root to follow TUF-1.0.9 section 5.1

erickt opened the pull request:
This switches Client::update_root to follow TUF-1.0.9 section 5.1 to update the root metadata. It now will fetch root N+1 until it receives a not-found error.

Test: This behavior is checked by the current tests. We could add a new test that makes sure we don't fetch `root.json`, but I'm not sure we get much value from a test like that.

erickt committed:
Change update_root to follow TUF-1.0.9 section 5.1

This switches Client::update_root to follow TUF-1.0.9 section 5.1 to update
the root metadata. It now will fetch root N+1 until it receives a not-found
error.

Change-Id: I79147f64575ae8c1aa3fa418990d44e11a7d506b

kellerb commented on src/client.rs:
```diff
@@ -463,67 +463,42 @@ where
-                .await?;
+            updated = true;
 
             if !self.tuf.update_root(&raw_signed_root)? {
```
can we change Tuf::update_root to require that the new root version == current root version +1?

erickt replied:
Yeah this was weird. I fixed that in another patch I'm working on, I'll pull it back here.

kellerb commented on src/client.rs:
```diff
@@ -463,67 +463,42 @@ where
+            //     the application using TUF. For example, W may be tens of kilobytes. The filename
+            //     used to download the root metadata file is of the fixed form
+            //     VERSION_NUMBER.FILENAME.EXT (e.g., 42.root.json). If this file is not available,
+            //     or we have downloaded more than Y number of root metadata files (because the
```
does this implementation mean rust-tuf is hardcoding Y to infinity?

erickt replied:
No, `Y` in our case is `self.config.max_root_length` down on line 484. By default it's 1 MiB.

kellerb replied:
`Y` is the number of new root metadata files to try downloading, i think `self.config.max_root_length` is `W`

erickt replied:
Oops, good catch. I filed #306 and added a FIXME to track that.

kellerb reviewed (changes requested):
For testing, since a key part of TUF is using the correct sequence of operations when interacting with untrusted repos, it seems like it would be good to have e.g. a RepositoryProvider type that allows us to verify the sequence of operations performed on it by Client, which would then make adding a test for this change easy.

Maybe this would be easy to make by creating a new RepositoryProvider that forwards to a wrapped EphemeralRepository after recording the method call.

erickt committed:
Address comments, add more testing

This adds a new test-only `TrackRepository` wrapper type, which allows
most interactions with repository storage and providers to be captured
and checked to make sure they are correct during testing.

Note: This removes the `versioned_init` tests, since it is now redundant
with the `chain_update_root` tests.

Change-Id: I14ca6764a4f29cdc42afcd7b701e655bc10a9f2a

erickt reviewed (commented):
You convinced me to add a helper repository that'll track changes to repository storage and providers.

Note: I removed the `versioned_init` tests, since that logic is now captured in the `root_chain_update` tests.

kellerb commented on src/repository/track_repo.rs:
```diff
@@ -0,0 +1,165 @@
+    }
+}
+
+/// Helper Repository wrapper that tracks all the fetches and stores for testing purposes.
```
"... tracks all the <ins>metadata</ins> fetches ..."

erickt replied:
Fixed, thanks.

kellerb commented on src/tuf.rs:
```diff
@@ -190,14 +189,12 @@ impl<D: DataInterchange> Tuf<D> {
+            let next_root_version = trusted_root
+                .version()
+                .checked_add(1)
+                .expect("root version should be less than max u32");
```
should we error instead of panic here?

semi-related, 2^32-1 doesn't seem very high for a process that could be automated

erickt replied:
Sure, I'll switch. We'd never trip over this in practice though because the new root would fail to deserialize a 2^32 value into a u32.

We could bump the version numbers up to a u64 are worthwhile, but consider the case where it only took 1 millisecond to create or fetch a root metadata. Trying to generate or fetch `(2^32-1) - N` would still take approximately 50 days. Repositories and clients should instead have separate monitoring in place to catch runaway processes like that.

kellerb reviewed (approved):
approved pending comments

erickt committed:
Address comments

* Fix a comment since TrackRepository only tracks metadata changes.
* Error rather than panic if root version is 2^32 or above.
* Add a fixme(#306) to add a limit on the number of root metadata fetch.

Change-Id: I8adef8261108ee45630c407738372fb37e93c3b4
